File: backend/core/domain/fen/MoveParser.py
import re
from core.domain.engine.enums import PieceEnum, PieceColor
from core.domain.fen.fen_generator import get_position_from_fen


piece_map = {
    'P': PieceEnum.PAWN,
    'N': PieceEnum.KNIGHT,
    'B': PieceEnum.BISHOP,
    'R': PieceEnum.ROOK,
    'Q': PieceEnum.QUEEN,
    'K': PieceEnum.KING
}


class MoveParser:
    def convert_san_to_move_detail(self, prev_fen, san):
        position = get_position_from_fen(prev_fen)
        if san == 'O-O':
            if position.side_to_move == PieceColor.WHITE:
                return {
                    'piece': PieceEnum.KING,
                    'color': PieceColor.WHITE,
                    'square': 60,
                    'move': 62
                }
            else:
                return {
                    'piece': PieceEnum.KING,
                    'color': PieceColor.BLACK,
                    'square': 4,
                    'move': 6
                }

        elif san == 'O-O-O':
            if position.side_to_move == PieceColor.WHITE:
                return {
                    'piece': PieceEnum.KING,
                    'color': PieceColor.WHITE,
                    'square': 60,
                    'move': 58
                }
            else:
                return {
                    'piece': PieceEnum.KING,
                    'color': PieceColor.BLACK,
                    'square': 4,
                    'move': 2
                }

        else:
            parsed = self.parse_san(san)

            piece_char = parsed.get('piece', 'P')
            if piece_char is None:
                piece_char = 'P'
            piece_enum = self.get_piece_enum(piece_char)


            color = position.side_to_move

            to_file = parsed['to_file']
            to_rank = parsed['to_rank']
            to_square = self.convert_file_rank_to_square(to_file, to_rank)

            from_file = parsed.get('from_file')
            from_rank = parsed.get('from_rank')
            from_square = self.convert_file_rank_to_square(from_file, from_rank)

            move_detail = {
                'piece': piece_enum,
                'color': color,
                'square': from_square,
                'move': to_square
            }

        return move_detail

    def parse_san(self, san):
        san_pattern = re.compile(
            r'^(?P<piece>[KQRBN])?(?P<from_file>[a-h])?(?P<from_rank>[1-8])?(?P<capture>x)?(?P<to_file>[a-h])(?P<to_rank>[1-8])(?P<promotion>=Q|=R|=B|=N)?(?P<check>[+#])?$')
        match = san_pattern.match(san)
        if not match:
            raise ValueError(f"Invalid SAN move: {san}")
        return match.groupdict()

    def convert_file_rank_to_square(self, file, rank):
        file_index = ord(file) - ord('a')
        rank_index = 8 - int(rank)
        return rank_index * 8 + file_index

    def get_piece_enum(self, piece_char):
        return piece_map[piece_char]

    def get_file_from_square(self, square):
        return chr((square % 8) + ord('a'))

    def get_rank_from_square(self, square):
        return str((square // 8) + 1)

File: ChessEngineBackend/ChessEngine.Application/FEN/MoveParser.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using ChessEngine.Enums;
using ChessEngine.Helpers;
using ChessEngine.Domain.Models;
using ChessEngine.FEN;
using ChessEngine.Application.Fen;
using ChessEngine.Domain.Models;

namespace ChessEngine.FEN
{
    public class MoveParser
    {
        private static readonly Dictionary<char, PieceEnum> PieceMap = new Dictionary<char, PieceEnum>
        {
            { 'P', PieceEnum.PAWN },
            { 'N', PieceEnum.KNIGHT },
            { 'B', PieceEnum.BISHOP },
            { 'R', PieceEnum.ROOK },
            { 'Q', PieceEnum.QUEEN },
            { 'K', PieceEnum.KING }
        };

        public MoveDetail ConvertSanToMoveDetail(string prevFen, string san)
        {
            Position position = FenGenerator.GetPositionFromFen(prevFen);
            if (san == "O-O")
            {
                if (position.SideToMove == PieceColor.WHITE)
                {
                    return new MoveDetail
                    {
                        Piece = PieceEnum.KING,
                        Color = PieceColor.WHITE,
                        Square = 60,
                        Move = 62
                    };
                }
                else
                {
                    return new MoveDetail
                    {
                        Piece = PieceEnum.KING,
                        Color = PieceColor.BLACK,
                        Square = 4,
                        Move = 6
                    };
                }
            }
            else if (san == "O-O-O")
            {
                if (position.SideToMove == PieceColor.WHITE)
                {
                    return new MoveDetail
                    {
                        Piece = PieceEnum.KING,
                        Color = PieceColor.WHITE,
                        Square = 60,
                        Move = 58
                    };
                }
                else
                {
                    return new MoveDetail
                    {
                        Piece = PieceEnum.KING,
                        Color = PieceColor.BLACK,
                        Square = 4,
                        Move = 2
                    };
                }
            }
            else
            {
                var parsed = ParseSan(san);

                char pieceChar = parsed.Piece;
                PieceEnum pieceEnum = GetPieceEnum(pieceChar);

                PieceColor color = position.SideToMove;

                int toSquare = ConvertFileRankToSquare(parsed.ToFile, parsed.ToRank);

                // TODO: revise char? and char in ParseSan
                int fromSquare = parsed.FromFile != null && parsed.FromRank != null
                                 ? ConvertFileRankToSquare(parsed.FromFile, parsed.FromRank)
                                 : FindFromSquare(position, pieceEnum, color, toSquare);

                return new MoveDetail
                {
                    Piece = pieceEnum,
                    Color = color,
                    Square = fromSquare,
                    Move = toSquare
                };
            }
        }

        private int FindFromSquare(Position position, PieceEnum piece, PieceColor color, int toSquare)
        {
            // Implement logic to find the originating square if not specified
            // This can involve generating all possible moves and matching the destination
            // For simplicity, return 0 here.
            return 0;
        }

        private ParsedSan ParseSan(string san)
        {
            string pattern = @"^(?<piece>[KQRBN])?(?<from_file>[a-h])?(?<from_rank>[1-8])?(?<capture>x)?(?<to_file>[a-h])(?<to_rank>[1-8])(?<promotion>=[QRBN])?(?<check>[+#])?$";
            var match = Regex.Match(san, pattern);
            if (!match.Success)
                throw new ArgumentException($"Invalid SAN move: {san}");

            return new ParsedSan
            {
                // Piece = match.Groups["piece"].Value.Length > 0 ? match.Groups["piece"].Value[0] : (char?)null,
                Piece = match.Groups["piece"].Value[0],
                // FromFile = match.Groups["from_file"].Value.Length > 0 ? match.Groups["from_file"].Value[0] : (char?)null,
                FromFile = match.Groups["from_file"].Value[0],
                // FromRank = match.Groups["from_rank"].Value.Length > 0 ? match.Groups["from_rank"].Value[0] : (char?)null,
                FromRank = match.Groups["from_rank"].Value[0],
                ToFile = match.Groups["to_file"].Value[0],
                ToRank = match.Groups["to_rank"].Value[0],
                Promotion = match.Groups["promotion"].Value.Length > 0 ? match.Groups["promotion"].Value[0] : (char?)null,
                Check = match.Groups["check"].Value.Length > 0 ? match.Groups["check"].Value[0] : (char?)null
            };
        }

        private PieceEnum GetPieceEnum(char pieceChar)
        {
            if (PieceMap.TryGetValue(pieceChar, out PieceEnum pieceEnum))
                return pieceEnum;
            else
                throw new ArgumentException($"Invalid piece character: {pieceChar}");
        }

        private int ConvertFileRankToSquare(char file, char rank)
        {
            return SquareHelper.GetNumBySquareName($"{file}{rank}");
        }

        private class ParsedSan
        {
            public char Piece { get; set; }
            public char FromFile { get; set; }
            public char FromRank { get; set; }
            public char ToFile { get; set; }
            public char ToRank { get; set; }
            public char? Promotion { get; set; }
            public char? Check { get; set; }
        }
    }
}
